compiler/runtime/OMRRSSReport.cpp

```diff
@@ -0,0 +1,354 @@
+            {
+            // Note: TR_VerboseLog will truncate the line to some maximum
+            sprintf(pagesInfo + pageCount * PAGE_PRINT_WIDTH, "%1zu ", resident);
+            }
+          else
+            {
+            TR_VerboseLog::writeLineLocked(TR_Vlog_PERF, "RSS: The following line is incomplete");
+            }
+         }
+#endif
+
+      TR_PersistentList<OMR::RSSItem> itemList = rssRegion->getRSSItemList(addr);
+
+      if (resident == 1 &&
+          itemList.isEmpty())
+         {
+         TR_VerboseLog::writeLineLocked(TR_Vlog_PERF, "RSS: Resident page at addr %p has no RSS items", addr);
+         }
+      else if (resident == 1)
+         {
+         ListIterator<OMR::RSSItem> it(&itemList);
+
+         size_t totalItemSize = 0;
+         int32_t countItems = 0;
+         uint8_t *prevEnd = (addr == start) ? addr : (uint8_t *)(((uintptr_t)addr / pageSize) * pageSize);  // current page start
+         OMR::RSSItem *prevItem = NULL;
+
+         uint64_t pageDebugCount = 0;
+
+         // print RSS items
+         for (OMR::RSSItem *item = it.getFirst(); item; item = it.getNext())
+            {
+            ListIterator<TR::DebugCounterAggregation> aggrIt(&item->_counters);
+            uint64_t itemDebugCount = 0;
+
+            // print debug counters
+            for (TR::DebugCounterAggregation *aggr = aggrIt.getFirst(); aggr; aggr = aggrIt.getNext())
+               {
+               uint64_t count = aggr->getCount();
+
+               itemDebugCount += count;
+               pageDebugCount += count;
+               if (count > 0) aggr->printCounters();
+               }
+
+            int32_t gap = item->_addr - prevEnd;
+
+            TR_ASSERT_FATAL(gap >= 0, "Item at addr %p in page %p is out of order, prevAddr=%p", item->_addr, addr,
+                            prevItem->_addr);
+
+            if (gap != 0)
+               {
+               TR_VerboseLog::writeLineLocked(TR_Vlog_PERF, "RSS: gap at addr=%p size=%zu in region %s", prevEnd, gap, rssRegion->_name);
+               }
+
+            totalItemSize += item->_size + gap;
+
+            TR_VerboseLog::writeLineLocked(TR_Vlog_PERF, "RSS item at addr=%p size=%zu itemDebugCount=%zu %s",
+                                           item->_addr, item->_size, itemDebugCount, OMR::RSSItem::itemNames[item->_type]);
+
+            countItems++;
+            prevEnd = item->_addr + item->_size;
+            prevItem = item;
+            }
+
+         TR_VerboseLog::writeLineLocked(TR_Vlog_PERF, "RSS: Page at addr %p has %zu bytes of items (%d items) pageDebugCount=%zu",
+                                        addr, totalItemSize, countItems, pageDebugCount);
+
+         TR_ASSERT_FATAL(totalItemSize <= pageSize, "Total size of items within page %p > page size\n", addr);
+         }
+
+      pageCount++;
+      }
+#endif
+
+   return rss;
+   }
+
+void
+OMR::RSSReport::printTitle()
+   {
+   int32_t regionNum = 0;
+   char rssLine[MAX_RSS_LINE];
```

Comment:
Wouldn't it be more readable if the regions were printed out sequentially, rather them being all on one line?

Reply:
If we want some frequent statistics it's much more readable in one line. For example, we might get this type of repport:

```
                                   cold cache cold cache
#PERF:  RSS Report(2 regions):   148(100%)  576(100%)  totalRSS=2896KB
#PERF:  RSS Report(2 regions):   148(100%)  576(100%)  totalRSS=2896KB
#PERF:  RSS Report(2 regions):   148(100%)  576(100%)  totalRSS=2896KB   
#PERF:  RSS Report(2 regions):     6(  4%)   72( 12%)  totalRSS=312KB   
#PERF:  RSS Report(2 regions):    35( 24%)  257( 45%)  totalRSS=1168KB 
#PERF:  RSS Report(2 regions):    45( 30%)  291( 50%)  totalRSS=1344KB
#PERF:  RSS Report(2 regions):    51( 34%)  306( 53%)  totalRSS=1428KB
#PERF:  RSS Report(2 regions):    54( 36%)  314( 54%)  totalRSS=1472KB
```
In other words, it's easier to see the changes in RSS for each region when printed in columns.

Reply:
Also please see https://github.com/eclipse/omr/pull/7349#issuecomment-2135770954